### PR TITLE
Converts StateAttribute hash from string to cached hash int keys

### PR DESCRIPTION
**Pull request by Kuranes**
Reducing further the GC work by not allocating nor concatenating strings each loop for shadergeneration when nothing does change

### CODE REVIEW

File: sources/osgShader/ShaderGenerator.js [hunk 1]
```diff
@@ -105,18 +107,17 @@ define( [
                     continue;
 
                 if ( attr.getHash ) {
-                    hash += attr.getHash();
+                    this.pushToHashStack( attr.getHash() );
                 } else {
-                    hash += attr.getType();
+                    this.pushToHashStack( attr.hashComputeCode( attr.getType() ) );
```
Comment (stephomi) [inline sources/osgShader/ShaderGenerator.js]: hashComputeCodeFromString

File: sources/osgShader/ShaderGenerator.js [hunk 2]
```diff
@@ -145,14 +146,13 @@ define( [
                         continue;
 
                     if ( attr.getHash ) {
-                        hash += attr.getHash();
+                        this.pushToHashStack( attr.getHash() );
                     } else {
-                        hash += attr.getType();
+                        this.pushToHashStack( attr.hashComputeCode( attr.getType() ) );
```
Comment (stephomi) [inline sources/osgShader/ShaderGenerator.js]: hashComputeCodeFromString

File: sources/osg/Utils.js [hunk 3]
```diff
@@ -226,5 +226,59 @@ define( [
...
+    Utils.hashComputeCodeFromIntList = function ( intList ) {
+
+        var hash = 0;
+        var intAt = 0;
+
+        if ( intList === undefined || intList.length === 0 ) {
+            return hash;
+        }
+
+        for ( var i = 0; i < intList.length; i++ ) {
```
Comment (stephomi) [inline sources/osg/Utils.js]: Always cache length...
Shouldn't it be intList.current ?
It's not a list but an array so maybe hashComputeCodeFromIntArray

File: sources/osg/Utils.js [hunk 4]
```diff
@@ -226,5 +226,59 @@ define( [
...
+    Utils.hashComputeCodeFromString = function ( str ) {
+
+        var hash = 0;
+        var chara = 0;
+
+        if ( str === undefined || str.length === 0 ) {
+            return hash;
+        }
+
+        for ( var i = 0; i < str.length; i++ ) {
```
Comment (stephomi) [inline sources/osg/Utils.js]: cache length

File: sources/osg/StateAttribute.js [hunk 5]
```diff
@@ -33,7 +37,9 @@ define( [
             return this.attributeType;
         },
 
-        apply: function () {},
+        apply: function () {
+            this.dirty( false );
```
Comment (stephomi) [inline sources/osg/StateAttribute.js]: I don't think I like that... we compute a code from a string (by iterating on a string) each time we apply an attribute.

We apply way more attribute than we generate a shader I think...

+ we have to make sure all the other apply will call the original stateAttribute apply (error prone?)

File: sources/osgShader/ShaderGenerator.js [hunk 6]
```diff
@@ -203,13 +215,17 @@ define( [
 
             return function ( state ) {
                 // extract valid attributes
-                var hash = '';
+
+                this._hashStack.current = 0;
```
Comment (stephomi) [inline sources/osgShader/ShaderGenerator.js]: It's bad practice to add attribute to an array.
Add an additional variable if needed.
Also... can the array be a global (global to all the ShaderGenerator)?

Anyway in that case I don't think it's super useful to avoid allocation...
We only push to this hash on a rendering leaf... (so typically a small number) and we only push a small of a stuffs (number of attributes).

I did some test a while ago by using a similar array workaround to avoid push/pop... and for relatively small array, I didn't notice any gain (probably because of the similar pushToHashStack and because js VM handles well small array).
But still, I'm against useless push/pop :).

File: sources/osg/Utils.js [hunk 7]
```diff
@@ -226,5 +226,59 @@ define( [
 
     } )();
 
+    Utils.hashComputeCodeFromString = function ( str ) {
+
+        var hash = 0;
+        var chara = 0;
+
+        if ( str === undefined || str.length === 0 ) {
```
Comment (stephomi) [inline sources/osg/Utils.js]: if ( !str ) do the same

File: sources/osg/Utils.js [hunk 8]
```diff
@@ -226,5 +226,59 @@ define( [
...
+        return hash;
+
+    };
+
+    Utils.hashComputeCodeFromIntList = function ( intList ) {
+
+        var hash = 0;
+        var intAt = 0;
+
+        if ( intList === undefined || intList.length === 0 ) {
```
Comment (stephomi) [inline sources/osg/Utils.js]: Maybe ``if ( !intList || intList.length === 0 )`` because if it's null ``intList.length`` won't fail but....

@cedricpinson Can we agree on ``!obj`` vs ``obj === undefined`` when we test an **object**?
I often made this comment but I don't know if we agreed or not on this practice

**Comment by stephomi** [comment]
Do you have a specific case/example where we win performances because I don't really think the hash building and test has ever been a bottleneck due to string concatenation (since it's only done per shader creation / draw leaf).

Also, we still do string concatenation because of the ``dirty( false )`` call in ``apply``

File: sources/osgShader/ShaderGenerator.js (hunk 1 above)
Comment (Kuranes) [inline sources/osgShader/ShaderGenerator.js]: indeed.
Yet I never ever encountered something going in this branch, and don't understand why it's there.

File: sources/osgShader/ShaderGenerator.js (hunk 6 above)
Comment (Kuranes) [inline sources/osgShader/ShaderGenerator.js]: Good point about that array member added. I think I checked how we did it in cullStack but maybe I misread.

Array could be global too, why not indeed.

Rationale and results are explained in the pivotal task, but anyway:

tl;dr: getHash does pop regularly on js profiler, this patch does reduce by half GC firefox in the boxe: https://sketchfab.com/models/663db593a73e4ceabf21a457f5ebcb26 and was my guess on the matcap/shadeless thing (as it's the only real difference between the two)

it's called each frame, very often (the cache flatten lots of not "flattened out" stateset like it theoritically could), and it'll it's all stateattributes which in turn does string concat of it parameter each frame + all texture attributes stack (notice the "stack")

File: sources/osg/StateAttribute.js (hunk 5 above)
Comment (stephomi) [inline sources/osg/StateAttribute.js]: setDirty( false ) ?

**Comment by Kuranes** [comment]
useful for diffing the rebase of death
